### PR TITLE
Fix `-Dc_std=...` not affecting cross compilation. Closes #5097. [0.50.* only]

### LINKED ISSUE #5097: 0.50 regression: Default option c_std not applied when cross-compiling
When providing `default_options: ['c_std=c99']` to the `project`, it is not being applied on `meson build --cross-file <cross-file>` invocation on a simple test project using flexible array members:

`ninja -v`

> [1/2] cc -Itest-executable@exe -I. -I../../work -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -g -pedantic -MD -MQ 'test-executable@exe/main.c.o' -MF 'test-executable@exe/main.c.o.d' -o 'test-executable@exe/main.c.o' -c ../../work/main.c
../../work/main.c:5:9: warning: ISO C90 does not support flexible array members [-Wpedantic]
     int array[];
         ^
[2/2] cc  -o test-executable 'test-executable@exe/main.c.o' -Wl,--no-undefined -Wl,--as-needed

However, after doing any kind of (even unrelated) configuration such as `meson configure -Db_asneeded=true`, the configuration is being fixed and applies as expected:

`ninja -v`

> [1/2] cc -Itest-executable@exe -I. -I../../work -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch **-std=c99** -g -pedantic -MD -MQ 'test-executable@exe/main.c.o' -MF 'test-executable@exe/main.c.o.d' -o 'test-executable@exe/main.c.o' -c ../../work/main.c
[2/2] cc  -o test-executable 'test-executable@exe/main.c.o' -Wl,--no-undefined -Wl,--as-needed

Note: I haven't tested whether this also applies to other default options.

---

#### main.c

```c
#include <stdio.h>

struct a {
    int foo;
    int array[];
};

int main(int argc, char* argv[]) {
    printf("Hey!\n");
    return 0;
}
```

#### meson.build

```meson
project('test-project', 'c', version: '0.0.1', default_options: ['c_std=c99'])
add_project_arguments('-pedantic', language: 'c')
executable('test-executable', 'main.c')
```

#### cross-file

```
[host_machine]
system = 'linux'
cpu_family = 'x86_64'
cpu = 'x86_64'
endian = 'little'

[properties]
root = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64'
needs_exe_wrapper = true

[paths]
prefix = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot'

[binaries]
c = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android28-clang'
cpp = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android28-clang++'
ar = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android-ar'
as = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android-as'
ld = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld'
strip = '/usr/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android-strip'
pkgconfig = '/usr/bin/pkg-config'
```

### PR DESCRIPTION
**Pull request by Ericson2314**
Will make a `0.50` hotfix cherry-pick if this looks good. CC @jpakkane @dcbaker @nirbheek.

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -5076,6 +5076,14 @@ def test_pkg_config_option(self):
         testdir = os.path.join(self.unit_test_dir, '55 pkg_config_path option')
         self.init(testdir, extra_args=['-Dpkg_config_path=' + os.path.join(testdir, 'extra_path')])
 
+    def test_std_remains(self):
+        # C_std defined in project options must be in effect also when cross compiling.
```
Comment (jpakkane) [inline run_unittests.py]: This should not say "also when cross compiling". This one tests native compilation, the other one is cross compilation

**Comment by jpakkane** [comment]
I don't like this as a proper solution as it is veering in the "hacks upon hacks" territory. For a point release this could be ok assuming the issue is fixed properly in master.

**Comment by Ericson2314** [comment]
Yes I agree this is hacky. Where we might differ is I think its happy because the current behavior is a bit arbitrary. I'll retarget this to 0.50, and make a new one for master with `build_` and `host_` across the board where appropriate.

I can do `0.50` the rebase this weekend for the hotfix release for sure, but not sure whether I can also do the `master` version before monday too, especially if https://github.com/mesonbuild/meson/pull/5173 still needs a test. Is that OK?

File: run_unittests.py (hunk 1 above)
Comment (Ericson2314) [reply run_unittests.py]: Thanks. Should be fixed.

**Comment by nirbheek** [comment]
Is this ready to go into 0.50?

**Comment by Ericson2314** [comment]
Yes!

**Pull request merged**